src/Nonlinear/parse.jl

```diff
@@ -46,7 +117,59 @@ function parse_expression(
     while !isempty(stack)
         parent_node, arg = pop!(stack)
         if arg isa MOI.ScalarNonlinearFunction
-            _parse_without_recursion_inner(stack, data, expr, arg, parent_node)
+            if haskey(data.cache, arg)
+                subexpr = data.cache[arg]
+                if subexpr isa Tuple{Expression,Int}
+                    _expr, _node = subexpr
+                    subexpr, I = _extract_subexpression!(data, _expr, _node)
+                    if expr === _expr
+                        if parent_node > first(I)
+                            @assert parent_node > last(I)
+                            parent_node -= length(I) - 1
+                        end
+                        for i in eachindex(stack)
+                            _parent_node = stack[i][1]
+                            if _parent_node > first(I)
+                                @assert _parent_node > last(I)
+                                stack[i] =
+                                    (_parent_node - length(I) + 1, stack[i][2])
+                            end
+                        end
+                    end
+                    for (key, val) in data.cache
+                        if val isa Tuple{Expression,Int}
+                            __expr, __node = val
+                            if _expr === __expr && __node > first(I)
+                                if __node <= last(I)
+                                    data.cache[key] = (
+                                        data.expressions[subexpr.value],
+                                        __node - first(I) + 1,
+                                    )
+                                else
+                                    data.cache[key] =
+                                        (__expr, __node - length(I) + 1)
+                                end
+                            end
+                        end
+                    end
+                    data.cache[arg] = subexpr
+                end
+                parse_expression(
+                    data,
+                    expr,
+                    subexpr::ExpressionIndex,
+                    parent_node,
+                )
+            else
+                _parse_without_recursion_inner(
+                    stack,
+                    data,
+                    expr,
+                    arg,
+                    parent_node,
+                )
+                data.cache[arg] = (expr, length(expr.nodes))
+            end
```

Comment:
So I think we can tidy up all of this, but I'm not opposed to this part necessarily. We could be cleverer in how we do this.

Reply:
Yes, this version should be working and tested so now it could be refactored :)